SETUP/db_schema.sql

```diff
@@ -423,6 +437,15 @@ CREATE TABLE `special_days` (
   UNIQUE KEY `spec_code` (`spec_code`)
 ) COMMENT='definitions of SPECIAL days';
 
+--
+-- Table structure for table `tally_snapshot_times`
+--
+
+CREATE TABLE `tally_snapshot_times` (
+  `timestamp` int(10) unsigned NOT NULL default '0',
```

Comment:
This assumes that all tallyboards are snapshotted on the same schedule. The current db design doesn't require that, but `take_tally_snaphots.php` assumes it, so it's probably safe?

Reply:
True. We could have it record a snapshot time for every Tallyboard but that seems excessive?

Reply:
> I want to keep each Tallyboard snapshot self-contained so I've kept the last snapshot timestamp as a column in `current_tallies`.

Hm.
- On the one hand, `current_tallies.last_snap_timestamp` suggests that, within a given TallyBoard, different holders could have different snapshot-times (which the TallyBoard code doesn't allow, and might cause some headaches if it ever did happen).
- On the other hand, the `tally_snapshot_times` table implies that all Tallyboards have the same snapshot times (which would violate the TallyBoard abstraction in the same way you object to above).

For an example of where the latter wouldn't work, consider a site that tweaked `take_tally_snapshots.php` to save some tallies nightly, and some weekly (but always at midnight). In the status quo, I think this would work. But with this PR, when that site calls `get_deltas` for a weekly tally (over some interval), it'll fetch all the daily snapshot times too, and set `$delta_at_` to 0 for all those other times. So the result will (incorrectly) look like a daily tally where (within each week) nothing happens for 6 days and then everything happens on one day.

Or consider a site where all TallyBoards are snapshotted daily, but some at noon and some at midnight. (This example seems less likely to me, but who knows.)

To resolve this, I suggest dropping `current_tallies.last_snap_timestamp`, and instead adding `tally_name` and `holder_type` columns to `tally_snapshot_times`, so each TallyBoard can have its own set of timestamps.

To get the latest timestamp for a given TallyBoard, we could just ask for `MAX(timestamp) WHERE ...`, which I'm guessing would be fairly quick, but could you also add an index on `(tally_name, holder_type, MAX(timestamp))`?

Reply:
Side note: I was looking at where we use time-of-latest-snapshot. `TallyBoard::get_info_from_latest_snapshot` fetches it and puts it into the array that it returns, but as far as I can tell, nobody uses that data. So that function could just not bother. [Not suggesting that for this PR.] On the other hand, maybe the recipients of that data *should* use the timestamp. (E.g., to say something like "as of $whenever server time".)

Reply:
> To resolve this, I suggest dropping `current_tallies.last_snap_timestamp`, and instead adding `tally_name` and `holder_type` columns to `tally_snapshot_times`, so each TallyBoard can have its own set of timestamps.

That makes sense to me too, I'll work on that.

Reply:
This is done. Note that it will take about 4 hours to populate `tally_snapshot_times` when we deploy this. Doing the `SELECT DISTINCT timestamp` for each tally_name/holder_type takes ~15 minutes because although it's an index scan, it takes that long just to (what I presume is) load the index from the disk to scan and I don't think the whole index will fit in memory.

Reply:
Bearing in mind my ignorance on this subject, does this imply that we should, at the very least, put out a general announcement to the site a day or two before the next code deployment that the site may/will be slow during "X" time frame? Or even have the site unavailable?

Reply:
My suggestion on the deployment plan would be:

After the DB backup happens and the tallyboards are snapshotted, we start the data-population script. Its likely to slow down the site just like a backup does but probably for longer -- there should be no need to make the site unavailable. After the data-population script finishes we could do the code deployment that uses the new tables.

Sometime later (a day or so) we could start the script to prune tally_delta=0 values from `past_tallies` but that isn't required for the new code to work and can happen slowly over several days.

After that script finishes we'll want to do an optimize table on `past_tallies` to actually shrink the table. I have no idea how long that will take. According to the [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/optimize-table.html) it can be done entirely online with only needing a table lock at the beginning and at the end. That plus given `past_tallies` queries are now only in the stats code, we probably don't need a downtime window for that either but should start it after the daily tallyboard snapshot code runs.